Fix ECH and version negotiation order. (#320)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

## Client-Facing Server

Upon receiving an "encrypted_client_hello" extension, the client-facing server

> determines if it can decrypt it, prior to negotiating any other TLS parameters.

Reviewer: s/can decrypt it/can accept ECH/, since "accept ECH" is a well-defined term.

Author: (Resolved by committing @chris-wood's suggestion.)

> determines if it can decrypt it, prior to negotiating any other TLS parameters.

Reviewer 2: ```suggestion
determines if it will accept ECH, prior to negotiating any other TLS parameters.
```

value of ServerHello.random[0:24] is generated as usual by invoking a secure
random number generator (see {{RFC8446}}, Section 4.1.2).

> \[\[TODO: If the ClientHello offers TLS 1.2 or below, it's possible the backend

Reviewer: Which ClientHello?

Reviewer 2: Presumably the inner, given below.

Author: Yeah, the backend server doesn't see the outer one.

server will negotiate it and leak the server name via an unencrypted server
certificate. This shouldn't happen because clients are forbidden from offering
earlier versions, but perhaps either the backend server should check (easy in

> Shared Mode, hard in Split Mode), or maybe the client-facing server should

Reviewer: Why is this hard in Split Mode? We could mandate that if the backend server observes the empty ECH extension (sent in the ClientHelloInner in order to signal the backend server to confirm acceptance), it aborts unless it can negotiate 1.3 or higher.
